src/stream/src/executor/source/mod.rs

```diff
@@ -41,3 +48,49 @@ pub async fn barrier_to_message_stream(mut rx: UnboundedReceiver<Barrier>) {
     }
     bail!("barrier reader closed unexpectedly");
 }
+
+pub fn get_split_offset_mapping_from_chunk(
+    chunk: &StreamChunk,
+    partition_idx: usize,
+    offset_idx: usize,
+) -> Option<HashMap<SplitId, String>> {
+    let mut split_offset_mapping = HashMap::new();
+    for (_, row) in chunk.rows() {
+        let split_id = row.datum_at(partition_idx).unwrap().into_utf8().into();
+        let offset = row.datum_at(offset_idx).unwrap().into_utf8();
+        split_offset_mapping.insert(split_id, offset.to_string());
+    }
+    Some(split_offset_mapping)
+}
```

Comment:
This is an additional round of iteration. We might need to benchmark.

Reply:
We also do it in parser, just moving the logic here. It is inevitable to iter over each chunk to get the latest offsets.